incubator/orm/auto_uint64.go

```diff
@@ -1,49 +1,216 @@
 package orm
 
 import (
+	"reflect"
+
 	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
 )
 
-func NewAutoUInt64TableBuilder(prefix byte, key sdk.StoreKey, cdc *codec.Codec) AutoUInt64TableBuilder {
-	return autoUInt64TableBuilder{prefix: prefix, key: key, cdc: cdc}
+const ormCodespace = "orm"
+
+var (
+	// todo: ormCodespace ok or do we need to claim error codes somehow?
+	ErrNotFound         = errors.Register(ormCodespace, 100, "not found")
+	ErrIteratorDone     = errors.Register(ormCodespace, 101, "iterator done")
+	ErrType             = errors.Register(ormCodespace, 102, "invalid type")
+	ErrUniqueConstraint = errors.Register(ormCodespace, 103, "unique constraint violation")
+)
+
+var _ TableBuilder = &autoUInt64TableBuilder{}
+
+func NewAutoUInt64TableBuilder(prefix []byte, key sdk.StoreKey, cdc *codec.Codec, model interface{}) *autoUInt64TableBuilder {
+	if len(prefix) == 0 {
+		panic("prefix must not be empty")
+	}
+	if cdc == nil {
+		panic("codec must not be empty")
+	}
+	if model == nil {
+		panic("model must not be empty")
+	}
+	tp := reflect.TypeOf(model)
+	if tp.Kind() == reflect.Ptr {
+		tp = tp.Elem()
+	}
+	return &autoUInt64TableBuilder{prefix: prefix, storeKey: key, cdc: cdc, model: tp}
 }
 
 type autoUInt64TableBuilder struct {
```

Comment:
Why is this a private struct? It basically forces people to refer to it as an interface outside of this package (TableBuilder). 

I try to go by the ["accept interfaces, return structs"](https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8) rule unless there is some exceptional case. I see no need not to return a public struct here. The caller can still treat it as an interface, but we don't need to throw away the type info

Reply:
I agree very much on the "accept interfaces, return structs" rule. Let's revisit all types and names after the functionality is okish.